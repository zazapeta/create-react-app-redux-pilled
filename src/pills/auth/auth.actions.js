import { createAction } from 'redux-actions';

import { addPrefixToActionTypes } from '../../redux-utils/utils';
import User from '../../models/user.model';

import { getMobilities } from '../mobilities/mobilities.actions';

export const ACTION_TYPES = addPrefixToActionTypes(
  {
    LOGIN_SUCCESS: 'LOGIN_SUCCESS',
    LOGIN_FAILURE: 'LOGIN_FAILURE',
    LOGOUT: 'LOGOUT',
  },
  'auth',
);

export const loginSuccess = createAction(ACTION_TYPES.LOGIN_SUCCESS);
export const loginFailure = createAction(ACTION_TYPES.LOGIN_FAILURE);
export const logout = createAction(ACTION_TYPES.LOGOUT);

export function login({ email, password, onPending, onSuccess, onFailure }) {
  onPending();
  return async (dispatch, getState, { api }) => {
    try {
      const rawUser = await api.auth.classicLogin({ email, password });
      const user = new User(rawUser);
      dispatch(loginSuccess(user));
      await dispatch(getMobilities()); // when the user login, we should fetch the mobilities as it is a part of the login process
      onSuccess(user);
    } catch (e) {
      dispatch(loginFailure(e));
      onFailure(e);
    }
  };
}
