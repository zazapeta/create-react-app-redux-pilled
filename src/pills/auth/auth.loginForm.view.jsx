import React from 'react';
import { reduxForm, Field } from 'redux-form';
import PropTypes from 'prop-types';

import { Button, Form } from '../../ui-kit';
import { Input } from '../../redux-form-utils/fieldComponents';
import t from '../../i18n';
import { required } from '../../redux-form-utils/fieldValidators';

function AuthView({ handleSubmit, invalid }) {
  return (
    <Form size="large" onSubmit={handleSubmit}>
      <Field
        name="email"
        component={Input}
        icon="user"
        iconPosition="left"
        placeholder="Email"
        validate={required}
        autoFocus
        fluid
      />
      <Field
        name="password"
        type="password"
        component={Input}
        icon="lock"
        iconPosition="left"
        placeholder={t('form.fields.password.label')}
        validate={required}
        fluid
      />
      <Button primary type="submit" fluid size="large" disabled={invalid}>
        {t('form.submit.login')}
      </Button>
    </Form>
  );
}

AuthView.propTypes = {
  onSubmit: PropTypes.func.isRequired,
};

export default reduxForm({
  form: 'AuthForm',
})(AuthView);
