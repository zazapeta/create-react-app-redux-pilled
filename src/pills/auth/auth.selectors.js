export const userSelector = ({ Auth }) => Auth.user;
