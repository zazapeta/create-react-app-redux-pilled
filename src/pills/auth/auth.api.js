import { authApi } from '../api/base.api';
import User from '../../models/user.model';
import { loadState } from '../../redux-utils/localStorage';

import { userSelector } from './auth.selectors';

/**
 * classic login as explained in the section 3.1 of the api doc
 * @param {User} - the user where to retrieve email and password field
 * @returns {Object} - return the raw data coming from the response object
 */
export async function classicLogin({
  email: USR_EMAIL,
  password: USR_PASSWORD,
}) {
  return authApi({ USR_EMAIL, USR_PASSWORD });
}

/**
 * remember me login as explained in the section 3.2 of the api doc
 * @param {User} - the user where to retrieve email and rememberMeId field
 * @returns {Object} - return the raw data coming from the response
 */
export async function rememberMeLogin({
  email: USR_REMEMBERME_EMAIL,
  rememberMeId: USR_REMEMBERME_ID,
}) {
  return authApi({ USR_REMEMBERME_EMAIL, USR_REMEMBERME_ID });
}

/**
 * after inscription login as explained in the section 3.3 of the api doc
 * @param {User} - the user where to retrieve remeberMeId and familyId field
 * @returns {Object} - return the raw data coming from the response
 */
export async function afterInscriptionLogin({
  rememberMeId: USR_REMEMBERME_ID,
  familyId: USR_REMEMBERME_FAMILY,
}) {
  return authApi({ USR_REMEMBERME_ID, USR_REMEMBERME_FAMILY });
}

/**
 * retrieve the user either:
 *  - from the current store stored in the local storage (synchronized with redux store)
 *  - as a result of a request (that may be classic, remberMe, afterInscription Login)
 * @param {fn} - function to use to do the login
 * @returns {User} - the current User as a an UserModel
 */
export async function getUser(login = classicLogin) {
  const rawUserFromState = userSelector(loadState());
  const userFromState = new User(rawUserFromState);
  const rawUser = await login(userFromState);
  return new User(rawUser);
}
