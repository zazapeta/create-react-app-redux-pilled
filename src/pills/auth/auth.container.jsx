import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import User from '../../models/user.model';

import { userSelector } from './auth.selectors';

const AuthContainer = ({ render, ...props }) => render(props);

function mapStateToProps(store) {
  return {
    user: userSelector(store),
  };
}

AuthContainer.propTypes = {
  user: PropTypes.instanceOf(User).isRequired,
  render: PropTypes.func.isRequired,
};

export default connect(mapStateToProps)(AuthContainer);
