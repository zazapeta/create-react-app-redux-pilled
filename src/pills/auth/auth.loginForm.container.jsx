import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  showErrorAlertModal,
  showBigLoaderModal,
  destroy,
} from '../modal/modal.actions';
import t from '../../i18n';

import AuthViewForm from './auth.loginForm.view';
import { login } from './auth.actions';

class AuthContainer extends Component {
  constructor(props) {
    super(props);
    this.handleLogin = this.handleLogin.bind(this);
  }

  handleLogin({ email, password }) {
    const { dispatch, onLogin } = this.props;
    const destroyModal = () => dispatch(destroy());
    dispatch(
      login({
        email,
        password,
        onPending: () =>
          dispatch(showBigLoaderModal({ content: t('modals.login_pending') })),
        onSuccess: (user) => {
          destroyModal();
          onLogin(user);
        },
        onFailure: () => {
          dispatch(
            showErrorAlertModal(
              {
                title: t('modals.login_error_title'),
                message: t('modals.login_error_message'),
                onClose: () => dispatch(destroy()),
              },
              dispatch,
            ),
          );
        },
      }),
    );
  }

  render() {
    return <AuthViewForm onSubmit={this.handleLogin} />;
  }
}

AuthContainer.propTypes = {
  onLogin: PropTypes.func.isRequired,
};

export default connect(null)(AuthContainer);
