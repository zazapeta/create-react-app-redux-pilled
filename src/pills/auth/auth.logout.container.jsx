import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { showConfirmModal, destroy } from '../modal/modal.actions';
import t from '../../i18n';

import { logout } from './auth.actions';

class LogoutContainer extends React.Component {
  constructor(props) {
    super(props);
    this.handleLogout = this.handleLogout.bind(this);
  }

  handleLogout() {
    const { dispatch } = this.props;
    dispatch(
      showConfirmModal(
        {
          title: t('modals.logout_confirm_title'),
          message: t('modals.logout_confirm_message'),
          onNo: () => dispatch(destroy()),
          onYes: () => dispatch(logout()),
        },
        dispatch,
      ),
    );
  }

  render() {
    return this.props.render({ logout: this.handleLogout });
  }
}

LogoutContainer.propTypes = {
  render: PropTypes.func.isRequired,
};

export default connect(null)(LogoutContainer);
