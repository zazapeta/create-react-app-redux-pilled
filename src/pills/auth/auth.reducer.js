import { ACTION_TYPES as AuthActionTypes } from './auth.actions';

const initialState = {
  user: {},
  error: {
    message: '',
  },
};

export default function AuthReducer(state = initialState, action) {
  const { type } = action;
  switch (type) {
    case AuthActionTypes.LOGIN_SUCCESS:
      return {
        ...state,
        user: action.payload,
        error: {
          message: '',
        },
      };
    case AuthActionTypes.LOGIN_FAILURE:
      return {
        ...state,
        user: {},
        error: {
          message: action.payload && action.payload.message,
        },
      };
    case AuthActionTypes.LOGOUT:
      return {
        ...state,
        user: {},
        error: {
          message: '',
        },
      };
    default:
      return state;
  }
}
