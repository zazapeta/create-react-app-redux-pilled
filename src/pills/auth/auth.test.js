import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import store from '../../redux-utils/store';
import LoginForm from './auth.loginForm.container';
import Logout from './auth.logout.container';
import { login, logout } from './auth.actions';
import AuthReducer from './auth.reducer';
import User from '../../models/user.model';
import { userSelector } from './auth.selectors';

const rawUser = {
  gSesGuid: '1C318AB4B890F07F3B4D574375DD2B3C',
  gUsrId: '1166',
  gUsrGuid: 'FFA6F5223A9955085D0A8E34EF3C9D15',
  gUsrLanguage: 'fr',
  gUsrDateFormat: 'dd/MM/yyyy',
  gUsrCompanyGuid: 'TOTAL',
  gServiceEnabled: '1',
  gCompanyMode: 'GLOBAL',
  gUsrEmail: 'test7@expateotest.com',
  gMixPanelToken: '124355cd44c781292f9f9f5bce580456',
  gTypeFormRpMain_FR: 'https://expateo.typeform.com/to/kkKgJU',
  gTypeFormRpMain_EN: 'https://expateo.typeform.com/to/XADq3N',
  gUsrDepartureCountryCode: 'MX',
  gRedirectPage: 'indexin.html#/Start',
  gRememberMeEmail: 'test7@expateotest.com',
  gRememberMeId: 'FFA6F5223A9955085D0A8E34EF3C9D15',
};

describe('Auth selectors', () => {
  it('Should return User', () => {
    let state = {
      Auth: {
        user: { gUsrId: '12345', gUsrGuid: '54321', gUsrLanguage: 'fr' },
      },
    };
    expect(userSelector(state)).toEqual({
      gUsrId: '12345',
      gUsrGuid: '54321',
      gUsrLanguage: 'fr',
    });
  });
});

describe('Auth render', () => {
  const noop = () => {};
  it('Should render', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <LoginForm onLogin={noop} />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should pass a logout function', (done) => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Logout
            render={({ logout }) => {
              expect(typeof logout).toEqual('function');
              done();
              return 'noop';
            }}
          />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Auth action - reducers', () => {
  it('LOGIN Should call pending & success', (done) => {
    let isOnPendingCalledFlag = false;
    let isOnSuccessCalledFlag = false;
    let isOnFailureCalledFlag = false;
    const loginAction = login({
      username: 'jhon',
      password: 'doe',
      onPending: () => (isOnPendingCalledFlag = true),
      onSuccess: () => (isOnSuccessCalledFlag = true),
      onFailure: () => (isOnFailureCalledFlag = true),
    });
    loginAction(store.dispatch, store.getState, {
      api: {
        auth: {
          classicLogin: () => Promise.resolve(rawUser),
        },
        mobilities: { getMobilities: () => Promise.resolve([]) },
      },
    })
      .then(() => {
        expect(isOnPendingCalledFlag).toEqual(true);
        expect(isOnSuccessCalledFlag).toEqual(true);
        expect(isOnFailureCalledFlag).toEqual(false);
        done();
      })
      .catch(done);
  });

  it('LOGIN Should call pending & failure', (done) => {
    let isOnPendingCalledFlag = false;
    let isOnSuccessCalledFlag = false;
    let isOnFailureCalledFlag = false;
    const loginAction = login({
      username: 'jhon',
      password: 'doe',
      onPending: () => (isOnPendingCalledFlag = true),
      onSuccess: () => (isOnSuccessCalledFlag = true),
      onFailure: () => (isOnFailureCalledFlag = true),
    });
    loginAction(store.dispatch, store.getState, {
      api: {
        auth: {
          classicLogin: () => Promise.reject(new Error()),
        },
      },
    })
      .then(() => {
        expect(isOnPendingCalledFlag).toEqual(true);
        expect(isOnSuccessCalledFlag).toEqual(false);
        expect(isOnFailureCalledFlag).toEqual(true);
        done();
      })
      .catch(done);
  });

  it('LOGIN:SUCCESS Should set the user', (done) => {
    const noop = () => {};
    const loginAction = login({
      username: 'jhon',
      password: 'doe',
      onPending: noop,
      onSuccess: noop,
      onFailure: noop,
    });
    AuthReducer(
      store.getState().Auth,
      loginAction(store.dispatch, store.getState, {
        api: {
          auth: {
            classicLogin: () => Promise.resolve(rawUser),
          },
          mobilities: { getMobilities: () => Promise.resolve([]) },
        },
      })
        .then(() => {
          expect(store.getState().Auth).toEqual({
            user: new User(rawUser),
            error: {
              message: '',
            },
          });
          done();
        })
        .catch(done),
    );
  });

  it('LOGIN:FAILURE Should set the error', (done) => {
    const noop = () => {};
    const loginAction = login({
      username: 'jhon',
      password: 'doe',
      onPending: noop,
      onSuccess: noop,
      onFailure: noop,
    });
    AuthReducer(
      store.getState().Auth,
      loginAction(store.dispatch, store.getState, {
        api: {
          auth: {
            classicLogin: () => Promise.reject(new Error('no way')),
          },
        },
      })
        .then(() => {
          expect(store.getState().Auth).toEqual({
            user: {},
            error: {
              message: 'no way',
            },
          });
          done();
        })
        .catch(done),
    );
  });

  it('LOGOUT: Should set reset the user and the error', () => {
    let state;
    state = AuthReducer(
      {
        user: { username: 'jhon-doe', firstname: 'doe' },
        error: {
          message: 'WAT',
        },
      },
      logout(),
    );
    expect(state).toEqual({
      user: {},
      error: {
        message: '',
      },
    });
  });
});
