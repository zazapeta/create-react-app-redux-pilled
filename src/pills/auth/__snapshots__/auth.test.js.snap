// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth render Should pass a logout function 1`] = `"noop"`;

exports[`Auth render Should render 1`] = `
<form
  className="ui large form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <div
      className="ui fluid left icon input"
    >
      <input
        autoFocus={true}
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <i
        aria-hidden="true"
        className="user icon"
      />
    </div>
  </div>
  <div
    className="field"
  >
    <div
      className="ui fluid left icon input"
    >
      <input
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Mot de passe"
        type="password"
        value=""
      />
      <i
        aria-hidden="true"
        className="lock icon"
      />
    </div>
  </div>
  <button
    className="ui large fluid primary disabled button"
    disabled={true}
    onClick={[Function]}
    role="button"
    tabIndex={-1}
    type="submit"
  >
    S'identifier
  </button>
</form>
`;
